<?xml version="1.0" encoding="UTF-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <head>
        <link rel="apple-touch-icon" sizes="57x57" href="../images/favicon/apple-icon-57x57.png">
        <link rel="apple-touch-icon" sizes="60x60" href="../images/favicon/apple-icon-60x60.png">
        <link rel="apple-touch-icon" sizes="72x72" href="../images/favicon/apple-icon-72x72.png">
        <link rel="apple-touch-icon" sizes="76x76" href="../images/favicon/apple-icon-76x76.png">
        <link rel="apple-touch-icon" sizes="114x114" href="../images/favicon/apple-icon-114x114.png">
        <link rel="apple-touch-icon" sizes="120x120" href="../images/favicon/apple-icon-120x120.png">
        <link rel="apple-touch-icon" sizes="144x144" href="../images/favicon/apple-icon-144x144.png">
        <link rel="apple-touch-icon" sizes="152x152" href="../images/favicon/apple-icon-152x152.png">
        <link rel="apple-touch-icon" sizes="180x180" href="../images/favicon/apple-icon-180x180.png">
        <link rel="icon" type="image/png" sizes="192x192" href="../images/favicon/android-icon-192x192.png">
        <link rel="icon" type="image/png" sizes="32x32" href="../images/favicon/favicon-32x32.png">
        <link rel="icon" type="image/png" sizes="96x96" href="../images/favicon/favicon-96x96.png">
        <link rel="icon" type="image/png" sizes="16x16" href="../images/favicon/favicon-16x16.png">
        <link rel="manifest" href="../images/favicon/manifest.json">
        <meta name="msapplication-TileColor" content="#ffffff">
        <meta name="msapplication-TileImage" content="/images/favicon/ms-icon-144x144.png">
        <meta name="theme-color" content="#ffffff">
        <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
        <link rel="stylesheet" type="text/css" href="../css/default.css" />
        <title>An Introduction to Variation Graphs urbanslug</title>
    </head>
    <body>
      <div id="wrapper">
        <header>
          <div id="header-content">
            <nav>
                <a href="../">Home</a>
                <a href="../notes.html">Notes</a>
                <a href="../about.html">About</a>
                <a href="../contact.html">Contact</a>
                <a href="../subscribe.html">Subscribe</a>
            </nav>
          </div><!--header-content-->
        </header>
        <article>
          <h1 id="post-title">An Introduction to Variation Graphs</h1>
          <div class="info">
    Posted on June 22, 2019
    
</div>
<div class="info">
    
    Tags: <a href="../tags/biology.html">biology</a>, <a href="../tags/variation%20graphs.html">variation graphs</a>, <a href="../tags/graphs.html">graphs</a>, <a href="../tags/bioinformatics.html">bioinformatics</a>, <a href="../tags/genetics.html">genetics</a>
    
</div>

<p>I expect the audience of this post to be both programmers and biologists so I’ll bring you up to speed on a topic before going into it. Feel free to skip a sentence, paragraph or even section if you’re familiar with a topic.</p>
<h1 id="genome-sequencing">Genome sequencing</h1>
<p>A <strong><a href="https://en.wikipedia.org/wiki/Genome">genome</a></strong> is the entire genetic code of an organism. While computational data is represented in binary form, ones, and zeros, biochemical data is represented by nitrogenous <a href="https://en.wikipedia.org/wiki/Base_(chemistry)">bases</a> that seem to stick out of a <a href="https://en.wikipedia.org/wiki/DNA">DNA</a> or <a href="https://en.wikipedia.org/wiki/RNA">RNA</a> molecule/strand abbreviated A, T, C, and G for DNA and A, G, C and U for RNA. We care about RNA because some viruses have RNA and not DNA.</p>
<p>If this is confusing, you can think of a strand of DNA or RNA as a thread with knots where each knot is a base.</p>
<p>A <strong>sequence</strong> is an <em>ordering</em> of something. A programming analog is a sequence vs a collection. Sequences are ordered, for example lists, and therefore have the potential for a <em>next</em> and a <em>previous</em> element while collections are just data thrown together, for example, a dictionary or a set.</p>
<p><strong>Genome sequencing</strong> (or sequencing a genome) therefore, is determining the order of bases in all of the DNA or RNA in an organism. What makes this easy is that all the cells in an individual organism have the same DNA so we can get all the DNA in an organism from a single cell. In practice, however, we can’t work with a single cell due to its size. Ignore chromosomes, haplotypes and other things you may know about DNA for now.</p>
<p>To determine the sequence of bases in an entire genome of an organism we focus on only one of the alleles (a <strong><a href="https://en.wikipedia.org/wiki/Haplotype">haplotype</a></strong>) and only one strand of the double helix. Since 2005 we’ve used methods broadly categorized under <strong>Next Generation Sequencing (NGS)</strong> to perform genome sequencing. There are two main ways of performing NGS:</p>
<ol style="list-style-type: decimal">
<li>Shear the DNA into small fragments, sequence those and try to build back the original sequence. An analogy that’s used is shredding a newspaper article then trying to recreate it.</li>
<li>Run the DNA strand like a train would run through a tunnel through a pore and read the sequence of bases. There are other methods of reading entire strands of DNA but they don’t matter in this context.</li>
</ol>
<p>As you would expect, each method has its drawbacks and advantages. What we get out of the machine that does the actual sequencing of DNA is called a read and reads have to be <a href="https://en.wikipedia.org/wiki/Sequence_alignment">aligned</a> and <a href="https://en.wikipedia.org/wiki/Sequence_assembly">assembled</a><sup>2</sup>. Alignment involves stacking reads on top of each other and assembling is the greater process that involves alignment, algorithmically choosing the best alignment and determining what the original sequence was.</p>
<p>There are two broad categories of assembly<sup>4</sup>:</p>
<ul>
<li><strong>De-novo assembly:</strong> this is where we sequence a genome that has never been sequenced before</li>
<li><strong>Mapping assembly:</strong> this is sequencing an organism’s unique code despite having the general sequence for the species. What you get from services like <a href="https://www.23andme.com/en-int/">23andMe</a>.</li>
</ul>
<h2 id="the-reference-genome">The reference genome</h2>
<p>A <a href="https://en.wikipedia.org/wiki/Reference_genome">reference genome</a> is a consensus sequence that accepted as the genome of a species<sup>2</sup>. It’s stored as one long sequence of characters/bases. You may wonder how we can have a known genome of an entire species when every individual has a unique genetic code or how <a href="https://www.scientificamerican.com/article/tiny-genetic-differences-between-humans-and-other-primates-pervade-the-genome/">humans are 99% chimp</a>. Well, the answer is that genetic code of most organisms is similar and this similarity increases as we narrow down taxonomically. When we say that <a href="https://www.scientificamerican.com/article/tiny-genetic-differences-between-humans-and-other-primates-pervade-the-genome/">a human is closer to a chimp than a monkey</a> what we mean is that we can observe greater variation between the genomes of the two, man+chimp vs monkey, than man vs chimp alone.</p>
<p>This isn’t actual math but may help clear things up.</p>
<pre><code>variation(combine_genomes(man, chimp), monkey) &gt; variation(man, chimp)
</code></pre>
<h1 id="variation-in-genomes">Variation in genomes</h1>
<p>However, there are still genomic differences and they should not be ignored. <em>The ignoring of differences is implicit in a linear reference.</em> A better way to describe them is to say that the differences are segregating within the population. We may also want to carry out a comparison between species or between related species which is done in <a href="https://en.wikipedia.org/wiki/Pan-genome">pangenomics</a>.</p>
<p>DNA has sections which are identical between individuals (conserved regions), and the number of these sections grow as we narrow down taxonomically and there are sections which vary between individuals, for example, the short sequence repeats that are compared in paternity testing.</p>
<p><a href="https://en.wikipedia.org/wiki/Graph_theory">Graph theory</a> is an area in math that can help us understand variable regions within genomes. The idea of representing genomes as graphs isn’t new, however, the low number of tools like <a href="https://github.com/vgteam/vg">vg</a> which apply graph theory to genomics and the little that we know about genomes has been a drawback.</p>
<h1 id="graphs">Graphs</h1>
<p>A <a href="https://en.wikipedia.org/wiki/Graph_theory">graph</a> is a series of vertices (also known as nodes) and edges. <img src="../images/Content/Graphs/all_graphs.png" alt="all graphs" /></p>
<p>For genome graphs, we focus on directed acyclic graphs. A <strong>walk</strong> in a directed graph is traversal from one node to another through an edge, for example, <em>a</em> to <em>b</em> to <em>d</em> or <em>a</em> to <em>c</em> to <em>d</em>. <img src="../images/Content/Graphs/directed_graph.png" alt="directed graph" /></p>
<h1 id="the-current-state-of-affairs">The current state of affairs</h1>
<p>Once the reference genome of an organism has been determined, it is stored in <a href="https://en.wikipedia.org/wiki/FASTA_format">fasta format</a> which contains the sequence and metadata. Moving forward, anyone sequencing the same species aligns against this reference. Differences that occur in less than 1% of the reads are usually thrown out; the ones that aren’t thrown out don’t help to update the reference but are stored in <a href="https://en.wikipedia.org/wiki/Variant_Call_Format">Variant Call Format (VCF)</a> which contains the variation data and their positions plus metadata. These VCF files are spread out amongst researchers and aid in the particular thing being researched but generally don’t contribute in and of themselves to the general genomic body of knowledge. However, every once in awhile the reference is updated but not on a fixed schedule<sup>2</sup>. It’s for this reason that the variation graph would be a good way of representing the reference. There is research that confirmed that short reads align better to the variation graph than to a linear reference<sup>3</sup>.</p>
<h1 id="graphs-and-genomes">Graphs and genomes</h1>
<p>Graphs that are applied to genomes are generally called <strong>genome graphs</strong>. However, there are two more specific categories which are sequence graphs and variation graphs.</p>
<p>As an example assume that we zoom on a hypothetical reference: <strong>“ACTGAATTTGTA”</strong></p>
<table>
<thead>
<tr class="header">
<th align="center">Variation</th>
<th align="center">Position</th>
<th align="center">Alternative</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">Variation1</td>
<td align="center">2</td>
<td align="center">GGGA</td>
</tr>
<tr class="even">
<td align="center">Variation2</td>
<td align="center">4</td>
<td align="center">C</td>
</tr>
</tbody>
</table>
<p>We could recursively insert Variation1 at position 2 and Variation2 at position 4 to generate the graph below:</p>
<p><img src="../images/Content/Graphs/example.png" alt="sequence graph" /> (generated using <a href="https://github.com/urbanslug/graphite">graphite</a> and <a href="https://github.com/urbanslug/graph">my current fork of graph</a>)</p>
<p>In this case, a single walk would represent a possible genome. Compared to the reference, this variation information is maintained and the graph still holds the data that was in reference.</p>
<h2 id="sequence-graphs">Sequence Graphs</h2>
<p>These are graphs with sequence labels on the nodes or edges.</p>
<p>Sequence graphs or equivalent structures have been used previously to represent multiple sequences that contain shared differences or ambiguities in a single structure. Related structures used in genome assembly which collapse long repeated sequences, so the same nodes are used for different regions of the genome include the <a href="https://en.wikipedia.org/wiki/De_Bruijn_graph">De Bruijn graph</a><sup>5</sup>. Graphs to represent genetic variation have previously been used for microbial genomes &amp; localized regions of the human genome such as the major <a href="https://en.wikipedia.org/wiki/Histocompatibility">histocompatibility</a> complex.</p>
<h2 id="variation-graphs">Variation Graphs</h2>
<p>A variation graph is a sequence graph together with a set of paths representing possible sequences from a population. However, what makes it so unique is it’s tight mapping between the graph and the reference.</p>
<h2 id="variation-graphs-and-rsv">Variation graphs and RSV</h2>
<p><a href="https://en.wikipedia.org/wiki/Human_orthopneumovirus">Human orthopneumovirus</a>, formerly known as Respiratory Syncytial Virus (RSV), is a single-stranded RNA virus and a good candidate for exploration using the variation graph because viruses don’t have proofreading in their genetic code. Proofreading is a process in which the cell ensures that it has copied the genetic code correctly in preparation for cell division. Without proofreading, errors will be commonplace leading to high mutation rates. Another advantage is the size of its genome; the reference stands at 15,206 bases which translate to 15206 bytes or 14.8 KB of memory.</p>
<h1 id="generating-a-variation-graph">Generating a variation graph</h1>
<p>As of writing this, <a href="https://github.com/urbanslug/graphite">graphite</a> can’t generate a graph out of reads alone (perform an alignment). It supports a reference in fasta and a single VCF file.</p>
<p>I’ll detail the algorithm in a <a href="../posts/2019-07-15-Creating-the-Initial-Variation-Graph.html">later post</a> but the gist of it is this:</p>
<ol style="list-style-type: decimal">
<li>Load the reference into memory or read a chunk of it if you wish</li>
<li>Load your variation data from a VCF</li>
<li>Organize variations into structs containing
<ul>
<li>variation</li>
<li>position</li>
<li>reference</li>
</ul></li>
<li>Sort the variations in ascending order by position</li>
<li>Using a right fold function - for support of streams
<ol style="list-style-type: decimal">
<li>fold through the list of variations</li>
<li>At each variation position split the reference and create a list of:
<ul>
<li>the string to the left</li>
<li>string to the right</li>
<li>a list of the variation and the base that was there originally (this will be a list of lists)</li>
</ul></li>
</ol></li>
<li>Create directed graph out of the list of lists generated by the fold
<ul>
<li><code>'((a b) (a c)) to become a node with edges from a to b and c to be and a-&gt;b and a-&gt;b</code></li>
</ul></li>
</ol>
<h2 id="variation">Variation</h2>
<p>A variation is a struct of <code>position</code> and <code>sequence</code>.</p>
<p>I’m using the <a href="https://github.com/stchang/graph">racket graph library graph</a> to generate a graph out of the nested lists and treating the graph as a “dynamic tree”.</p>
<p>We then rely on graph to generate an unweighted directed graph through <a href="https://docs.racket-lang.org/graph/index.html#%28def._%28%28lib._graph%2Fmain..rkt%29._unweighted-graph%2Fdirected%29%29">unweighted-graph/directed</a>. We export the graph in dot format and visualize via <a href="https://en.wikipedia.org/wiki/Graphviz">graphviz</a>. Serialization isn’t implemented yet.</p>
<h1 id="references">References</h1>
<ol style="list-style-type: decimal">
<li>Adam M. Novak, Erik Garrison, Benedict Paten A graph extension of the positional burrows-wheeler transform and its applications bioRxiv 051409; doi: https://doi.org/10.1101/051409</li>
<li>Church DM, Schneider VA, Graves T, Auger K, Cunningham F, Bouk N, et al. (2011) Modernizing Reference Genome Assemblies. PLoS Biol 9(7): e1001091. https://doi.org/10.1371/journal.pbio.1001091</li>
<li>Garrison, Erik &amp; Sirén, Jouni &amp; M Novak, Adam &amp; Hickey et al. (2018). Variation graph toolkit improves read mapping by representing genetic variation in the reference. Nature Biotechnology. 36. 10.1038/nbt.4227</li>
<li>Wolf, Beat. “De novo genome assembly versus mapping to a reference genome” (PDF). University of Applied Sciences Western Switzerland. Retrieved 6 April 2019.</li>
<li>Holley, Guillaume &amp; Peterlongo, Pierre. (2012). BLASTGRAPH: Intensive approximate pattern matching in sequence graphs and de-Bruijn graphs. Proceedings of the Prague Stringology Conference, PSC 2012.</li>
</ol>

<script src="https://carnivalapp.io/sites/413/init.js"></script>
<script>
  document.addEventListener("DOMContentLoaded", Carnival.init);
</script>

        </article>

        <footer>
  <p>All work is under the terms of the <a href="http://www.gnu.org/copyleft/fdl.html">GNU Free Documentation License, Version 1.3</a></p>
  <p>Generated with <a href="https://jaspervdj.be/hakyll/">Hakyll</a>.</p>
        </footer>
      </div> <!-- #wrapper -->
    </body>
</html>
